Lab 4.1

4.1.1
Hands-on DH
1. Select and agree with your partner on a 5-digit prime number, p.
2. Select and agree on a primitive-root of your prime number, g. 
You can get a list of prime number and find their primative roots at http://www.bluetulip.org/2014/programs/primitive.html
3. Each of you should independently select a secret integer. Let’s call that s. Keep them fairly small to avoid overflows if using a calculator.
4. Each of you independently calculate g raised to your secret integer, s, then calculate that value mod p  (g^s mod p).
5. Share your resulting value with your partner.
6. Each of you independently calculate the value of your partner’s result raised to your secret integer s, mod p. This is the shared secret, K.

In [14]:
# Function to calculate the greatest common divisor (gcd) of two numbers using the Euclidean algorithm
# def gcd(a, b):
#     while b:
#         a, b = b, a % b
#     return a

# Function to check if a number is a primitive root for a given prime number
def is_primitive_root(root, prime):
    powers = set()
    # Iterate through powers from 1 to (prime - 2)
    for i in range(1, prime - 1):
        # Calculate (root ^ i) % prime
        power = pow(root, i, prime)
        # If the result is already in the set, this is not a primitive root
        if power in powers:
            return False
        # Add the result to the set
        powers.add(power)
    # If all results are unique, this is a primitive root
    return True

# Function to find all primitive roots of a given prime number
def find_primitive_roots(prime):
    primitive_roots = []
    # Iterate through numbers from 2 to (prime - 1)
    for candidate_root in range(2, prime):
        # Check if the current number is a primitive root
        if is_primitive_root(candidate_root, prime):
            # Add the primitive root to the list
            primitive_roots.append(candidate_root)
    return primitive_roots

# Main function to run the program
def main():
    # Get the prime number as input
    prime = int(input("Enter a prime number: "))
    # Find all primitive roots for the given prime number
    primitive_roots = find_primitive_roots(prime)
    # Calculate the smallest primitive root using the min() function
    smallest_primitive_root = min(primitive_roots)
    
    # Print the list of all primitive roots
    print(f"All primitive roots of {prime}: {primitive_roots}")
    # Print the smallest primitive root
    print(f"The smallest primitive root of {prime} is {smallest_primitive_root}")

# Execute the main function
if __name__ == "__main__":
    main()


All primitive roots of 10007: [5, 7, 10, 14, 15, 17, 20, 21, 23, 28, 29, 30, 31, 34, 37, 40, 41, 42, 43, 45, 46, 51, 55, 56, 58, 59, 60, 62, 63, 65, 67, 68, 69, 74, 77, 80, 82, 84, 86, 87, 90, 91, 92, 93, 95, 101, 102, 103, 107, 110, 111, 112, 116, 118, 120, 123, 124, 125, 126, 129, 130, 133, 134, 135, 136, 137, 138, 148, 153, 154, 157, 160, 163, 164, 165, 167, 168, 172, 173, 174, 175, 177, 180, 182, 184, 186, 187, 189, 190, 191, 193, 195, 199, 201, 202, 204, 206, 207, 214, 220, 221, 222, 224, 227, 229, 231, 232, 233, 235, 236, 240, 241, 245, 246, 248, 250, 251, 252, 253, 258, 260, 261, 263, 265, 266, 268, 270, 272, 273, 274, 276, 277, 279, 285, 293, 296, 299, 303, 305, 306, 307, 308, 309, 314, 317, 319, 320, 321, 323, 326, 328, 329, 330, 331, 333, 334, 336, 341, 343, 344, 346, 348, 349, 350, 353, 354, 355, 360, 364, 365, 367, 368, 369, 371, 372, 374, 375, 377, 378, 380, 382, 383, 386, 387, 389, 390, 395, 398, 399, 402, 403, 404, 405, 407, 408, 411, 412, 414, 415, 425, 427, 428, 437, 4

In [32]:
# calculate g raised to your secret integer, s, then calculate that value mod p  (g^s mod p)
5**4 % 10007

625

In [33]:
# calculate g raised to your secret integer, s, then calculate that value mod p  (g^s mod p)
5**3 % 10007

125

In [38]:
# Alice
K_A1 = 125**4 % 10007

# Bob
K_B1 = 625**3 % 10007

print(K_A1)
print(K_B1)

if K_A1 == K_B1:
    print(f"SUCCESS -> We have a shared secrete key (K) = ({K_A1})")

9853
9853
SUCCESS -> We have a shared secrete key (K) = (9853)


1. 10007
2. 5
3. 4, 3 (unique secret integer 's')
4. g^s mod p = 5**4 % 10009
5. 625, 125
6. 9853

What value did each of you get for K? - We both got the same K (9853)
What were your p and q? - 10007, 5
What were your secret integers? - 4,3

4.1.2
Try this:

Assume Alice and Bob agreed on p=53, g=8.
Alice and Bob select secret integers, a and b.
Alice sent the value 48 to Bob.
Bob sent the value 35 to Alice.
Could you find K? What was it?
You would need to solve: (35^a mod 53 = 48^b mod 53)
How did you get your answer?

Agree on public parameters
p = 53
g = 8

In [19]:
# run the main function to verify p and g are accurate.
main()

All primitive roots of 53: [2, 3, 5, 8, 12, 14, 18, 19, 20, 21, 22, 26, 27, 31, 32, 33, 34, 35, 39, 41, 45, 48, 50, 51]
The smallest primitive root of 53 is 2


Generate private keys and compute public keys

In [20]:
# Alice -> generate public key
(8**4) % 53

15

In [21]:
# Bob -> generate public key
(8**3) % 53

35

Calculate shared secret key

In [25]:
# Alice -> calculate shared secret key
K_A = (35**4) % 53
print(K_A)

36


In [26]:
# Bob -> calculate shared secret key
K_B = (15**3) % 53
print(K_B)

36


In [28]:
# verify again that the K is the same
if K_A == K_B:
    print(True)

True


The K (key) above is a shared secret key and should be the same.

In [ ]:
""" 
The Diffie-Hellman key exchange is an algorithm that allows two parties to establish a shared secret key - 
over an insecure communication channel.

Below are the steps of the Diffie-Hellman key exchange:

1. Agree on public parameters (p, g)
    Alice                   Public Channel                  Bob
     [p, g] ------------------------------------------------> [p, g]
     [p, g] <------------------------------------------------ [p, g]

2. Generate private keys (a, b) and compute public keys (A, B)
    Alice                   Public Channel                  Bob
     [a]                                                   [b]
     [A = g^a mod p] -------> Public Key Exchange ------> [B = g^b mod p]

3. Calculate shared secret key (K)
    Alice                                               Bob
     [K_A = B^a mod p]                                    [K_B = A^b mod p]

4. Verify that K_A = K_B (Both parties now have the shared secret key K)
    Alice                                               Bob
     [K]                                                  [K]
"""
